tinygrad/runtime/ops_qcom.py

```diff
@@ -0,0 +1,310 @@
+        | ((local_size[2] - 1) << adreno.A6XX_HLSQ_CS_NDRANGE_0_LOCALSIZEZ__SHIFT),
+        global_size_mp[0], 0, global_size_mp[1], 0, global_size_mp[2], 0, # global size x,y,z followed by offsets
+        0xccc0cf, 0x2fc,
+        global_size[0], global_size[1], global_size[2], # original global sizes
+    )
+    self.reg(
+      adreno.REG_A6XX_SP_CS_CTRL_REG0,
+      (adreno.THREAD128 << adreno.A6XX_SP_CS_CTRL_REG0_THREADSIZE__SHIFT)
+      | (prg.halfreg << adreno.A6XX_SP_CS_CTRL_REG0_HALFREGFOOTPRINT__SHIFT)
+      | (prg.fullreg << adreno.A6XX_SP_CS_CTRL_REG0_FULLREGFOOTPRINT__SHIFT)
+      | (4 << adreno.A6XX_SP_CS_CTRL_REG0_BRANCHSTACK__SHIFT),
+      0x41, 0, prg.prg_offset, # offsets
+      *data64_le(prg.lib_gpu.va_addr), 0, *data64_le(prg.private_gpu.va_addr), prg.private_gpu.size,
+    )
+    self.cmd(adreno.CP_LOAD_STATE6_FRAG,
+             (adreno.ST_CONSTANTS << adreno.CP_LOAD_STATE6_0_STATE_TYPE__SHIFT)
+             | (adreno.SS6_INDIRECT << adreno.CP_LOAD_STATE6_0_STATE_SRC__SHIFT)
+             | (adreno.SB6_CS_SHADER << adreno.CP_LOAD_STATE6_0_STATE_BLOCK__SHIFT)
+             | (256 << adreno.CP_LOAD_STATE6_0_NUM_UNIT__SHIFT),
+             *data64_le(kernargs))
+    self.cmd(adreno.CP_LOAD_STATE6_FRAG,
+            (adreno.ST_SHADER << adreno.CP_LOAD_STATE6_0_STATE_TYPE__SHIFT)
+            | (adreno.SS6_INDIRECT << adreno.CP_LOAD_STATE6_0_STATE_SRC__SHIFT)
+            | (adreno.SB6_CS_SHADER << adreno.CP_LOAD_STATE6_0_STATE_BLOCK__SHIFT)
+            | (math.ceil(prg.image_size / 128) << adreno.CP_LOAD_STATE6_0_NUM_UNIT__SHIFT),
+            *data64_le(prg.lib_gpu.va_addr))
+    self.reg(adreno.REG_A6XX_SP_CS_INSTRLEN, prg.image_size // 4)
+    self.cmd(adreno.CP_RUN_OPENCL, 0)
+    self.cmd(adreno.CP_WAIT_FOR_IDLE)
+
+  def _update_exec(self, cmd_idx, global_size, local_size):
+    if global_size is not None:
+      self._patch(cmd_idx, offset=11, data=global_size)
+      self.cmd_idx_to_dims[cmd_idx][0] = global_size
+
+    if local_size is not None:
+      payload = (3 | ((local_size[0] - 1) << adreno.A6XX_HLSQ_CS_NDRANGE_0_LOCALSIZEX__SHIFT)
+        | ((local_size[1] - 1) << adreno.A6XX_HLSQ_CS_NDRANGE_0_LOCALSIZEY__SHIFT)
+        | ((local_size[2] - 1) << adreno.A6XX_HLSQ_CS_NDRANGE_0_LOCALSIZEZ__SHIFT))
+
+      self._patch(cmd_idx, offset=2, data=[payload])
+      self.cmd_idx_to_dims[cmd_idx][1] = local_size
+
+    global_size_mp = self.cmd_idx_to_dims[cmd_idx][0]
+    if self.cmd_idx_to_dims[cmd_idx][1]:
+      global_size_mp = cast(Tuple[int,int,int], tuple(int(g*l)for g,l in zip(
+        self.cmd_idx_to_dims[cmd_idx][0], self.cmd_idx_to_dims[cmd_idx][1])))
+    self._patch(cmd_idx, offset=3, data=[global_size_mp[0], 0, global_size_mp[1], 0, global_size_mp[2], 0])
+
+class QcomProgram(HCQProgram):
+  def __init__(self, device: QcomDevice, name: str, lib: bytes):
+    self.device, self.name, self.lib = device, name, lib
+
+    self.private_gpu = self.device._gpu_alloc(0x101, 0xC0F00)
```

Comment:
private mem should be shared for all programs, no?

Reply:
yeah, just noticed it too. done.